systemdspawner/systemd.py

```diff
@@ -76,73 +76,75 @@ async def start_transient_service(
-    # Can be overridden in unit_extra_properties.
-    properties.setdefault("OOMPolicy", "continue")
-
-    # ensure there is a runtime directory where we can put our env file
-    # If already set, can be space-separated list of paths
-    runtime_directories = properties.setdefault("RuntimeDirectory", unit_name).split()
-
-    # runtime directories are always resolved relative to `/run`
-    # grab the first item, if more than one
-    runtime_dir = os.path.join(RUN_ROOT, runtime_directories[0])
-    # make runtime directories private by default
+    properties = (properties or {}).copy()
+
+    # Ensure there is a runtime directory where we can put our env file, make
+    # runtime directories private by default, and preserve runtime directories
+    # across restarts to allow `systemctl restart` to load the env.
+    #
+    # ref: https://www.freedesktop.org/software/systemd/man/systemd.exec.html#RuntimeDirectory=
+    # ref: https://www.freedesktop.org/software/systemd/man/systemd.exec.html#RuntimeDirectoryMode=
+    # ref: https://www.freedesktop.org/software/systemd/man/systemd.exec.html#RuntimeDirectoryPreserve=
+    #
+    properties.setdefault("RuntimeDirectory", unit_name)
     properties.setdefault("RuntimeDirectoryMode", "700")
-    # preserve runtime directories across restarts
-    # allows `systemctl restart` to load the env
     properties.setdefault("RuntimeDirectoryPreserve", "restart")
 
-    if properties:
-        for key, value in properties.items():
-            if isinstance(value, list):
-                run_cmd += [f"--property={key}={v}" for v in value]
-            else:
-                # A string!
-                run_cmd.append(f"--property={key}={value}")
+    # Ensure that out of memory killing of a process run inside the user server
+    # (systemd unit), such as a Jupyter kernel, doesn't result in stopping or
+    # killing the user server.
+    #
+    # ref: https://www.freedesktop.org/software/systemd/man/systemd.service.html#OOMPolicy=
+    #
+    properties.setdefault("OOMPolicy", "continue")
 
+    # Pass configured properties via systemd-run's --property flag
+    for key, value in properties.items():
+        if isinstance(value, list):
+            # The properties dictionary is allowed to have a list of values for
+            # each of its keys as a way of allowing the same key to be passed
+            # multiple times.
+            run_cmd += [f"--property={key}={v}" for v in value]
+        else:
+            # A string!
+            run_cmd.append(f"--property={key}={value}")
+
+    # Create and reference an environment variable file in the first
+    # RuntimeDirectory entry, which is a whitespace-separated list of directory
+    # names.
+    #
+    # ref: https://www.freedesktop.org/software/systemd/man/systemd.exec.html#EnvironmentFile=
+    #
     if environment_variables:
+        runtime_dir = os.path.join(RUN_ROOT, properties["RuntimeDirectory"].split()[0])
```

Comment:
When refactoring this, just drop the `RUN_ROOT` for `/run` in the context of system services this will never change and always be `/run`. Also, the runtime directory shouldn't need to be exposed manually, since systemd will always define `RUNTIME_DIRECTORY` for services using the `RuntimeDirectory=` property.

Reply:
Thank you @behrmann for working on this project and chiming in!!!!

I think its fine to refactoring away `RUN_ROOT` - something I was inclined to do myself but didn't to keep things focused. 

> Also, the runtime directory shouldn't need to be exposed manually, since systemd will always define `RUNTIME_DIRECTORY` for services using the `RuntimeDirectory=` property.

Hmmm this is something I don't understand so well. Do you want to create a PR about these things?

Reply:
> Hmmm this is something I don't understand so well. Do you want to create a PR about these things?

The environment variables that should be defined are written to a file so that the token doesn't leak via the commandline. What I'm saying is that a variable for the runtime directory doesn't need to be passed via this file, since systemd will always set a variable `RUNTIME_DIRECTORY` when the property `RuntimeDirectory=` is used. 

As a side note. Passing secrets via environment variables is something that's really frowned upon (since the secrets still leak via `/proc/$PID/environ` to all child processes), which is why systemd added the credentials feature, which I use in #100, but the lower version bound is still a bit too low for that. An option to read the token form a file instead of taking it via an environment variable would be a great addition for the singleuser servers.